### DISCUSSION TITLE
[BP#65] Mettre en cache le Bytecode code intermédiaire

### DISCUSSION THREAD

**Discussion by devcrafting** [cnumr/best-practices#215] [Bonnes pratiques] [answered]
Vraiment encore d'actualité ? Ce n'est pas défaut avec les dernières versions de PHP?

Spécifique à PHP?

**Comment by ACTLEM** [comment]
Pour moi c'est embarqué par défaut (mais je ne suis pas spécialiste de PHP). @arnolem @yannkozon et autres spécialistes PHP, est-ce qu'il y a encore beaucoup d'endroits où ce n'est pas activé par défaut?

**Reply by arnolem** [reply]
Oui, c'est intégré à PHP, il n'y a rien à faire pour avoir un opCache. 👍

(Ca reste à configurer pour faire du tuning sur des grosses applications mais la majorité des devs n'ont pas à s'en soucier)
Le sujet est donc plutôt d'encourager à utiliser des versions récentes de PHP qui ont des gains significatifs de performance.

**Reply by ACTLEM** [reply]
Merci. Donc si on garde cette BP même si elle est propre à PHP (mais pourquoi pas étant donné la forte utilisation sur le web), l'idée serait plutôt de recommander d'avoir une version récente (> à x?) que de parler uniquement d'opCache, non?

**Reply by arnolem** [reply]
Oui, c'est ça. 
Il n'y a pas de version de PHP à recommander car la nouvelle est systématiquement plus performante que la précédente (même si on est en train d'atteindre une limite).
Depuis PHP 5.5, l'opcache est intégré par défaut. Il faut toutefois le configurer si on veut optimiser la gestion du cache.
PHP 7.0 apporte un gros gain de performance, ainsi que les versions intermédiaire 7.1, 7.2, 7.3, ...
PHP 8.0 arrive avec le JIT pour être encore plus performant dans des cas d'usage spécifique.
PHP 8.1 devrait également apporter des optimisations de performance (Je pense à fibers qui va permettre l'optimisation de librairies qui font de l'asynchrone). Même si ce n'est pas l'atout principale de cette version, la performance va toujours en s'améliorant depuis plus de 10 ans.

Donc la BP est obsolète mais l'encouragement à utiliser des versions plus récentes de PHP prend beaucoup plus de sens (Opcache, jit, ...)

**Reply by tbroyer** [reply]
Dans ce cas est-ce qu'on ne pourrait pas généraliser ? Ruby aussi a gagné un compilateur JIT ces dernières années.

**Reply by ACTLEM** [reply]
@arnolem @tbroyer si on généralise cette BP (ou pas), vous proposeriez quoi?

**Reply by tbroyer** [reply]
L'idée serait de s'assurer qu'on utilise en règle générale la version la plus récente du "langage" (runtime ?) pour bénéficier des gains en performance qu'elle ne manquerait pas d'apporter, ou a minima une version avec un compilateur juste-à-temps.

**Reply by rlemaire** [reply]
J'aime beaucoup l'idée d'utiliser la version la plus récente de la stack coté back. Ca marche aussi avec node ou java.

Par contre ça n'a plus rien à voir avec la BP d'origine.

**Comment by ACTLEM** [comment, marked as answer]
J'ai créé deux issues:

- Suppression de la 65: https://github.com/cnumr/best-practices/issues/280
- Ajout: Utiliser la version la plus récente du langage (affectée à @arnolem à l'origine de l'idée ;-) ): https://github.com/cnumr/best-practices/issues/279
